Support xpu host allocator

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123080

# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

guangyey committed:
support xpu host allocator

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

EikanWang commented on aten/src/ATen/xpu/CachingHostAllocator.cpp:
```diff
@@ -0,0 +1,77 @@
+
+struct XPUHostAllocator final
+    : public HostAllocatorInterface<XPUHostAllocatorImpl> {
+  at::DataPtr allocate(size_t size) override {
+    auto ptr_and_ctx = impl_->allocate(size);
+    return {
+        ptr_and_ctx.first,
+        ptr_and_ctx.second,
+        &raw_local_deleter,
+        at::DeviceType::CPU};
+  }
```
There is no difference between CUDA and XPU. Is it possible to move the function to the parent class?

guangyey replied:
`raw_local_deleter` is a deleter, which is necessary to define by an instance `caching_host_allocator`. So we need a forward declaration. If we move this function to the parent class,  we have to define `raw_local_deleter` under namespace `at` for each backend. There will be duplicated definitions for `raw_local_deleter` under the same namespace `at` even though the current PyTorch is the exclusive backend. It is not elegant.

EikanWang commented on aten/src/ATen/xpu/CachingHostAllocator.cpp:
```diff
@@ -0,0 +1,77 @@
+  }
+};
+
+static XPUHostAllocator host_allocator;
```
Should we move `host_allocator` to the anonymous namespace?

guangyey replied:
It is ok to move to an anonymous namespace. Done.

EikanWang commented on torch/_C/__init__.pyi.in:
```diff
@@ -1916,6 +1916,7 @@ def _xpu_getCurrentStream(device: _int) -> Tuple: ...
 def _xpu_getCurrentRawStream(device: _int) -> _int: ...
 def _xpu_synchronize(device: _int) -> None: ...
 def _xpu_emptyCache() -> None: ...
+def _xpu_xpuHostAllocator() -> _int: ...
```
```suggestion
def _xpu_HostAllocator() -> _int: ...
```

guangyey replied:
done.

EikanWang commented on torch/csrc/xpu/Module.cpp:
```diff
@@ -335,6 +343,10 @@ static struct PyMethodDef _THXPModule_methods[] = {
      METH_VARARGS | METH_KEYWORDS,
      nullptr},
     {"_xpu_synchronize", THXPModule_xpuSynchronize, METH_O, nullptr},
+    {"_xpu_xpuHostAllocator",
```
```suggestion
    {"_xpu_HostAllocator",
```

guangyey replied:
done.

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support xpu host allocator"




[ghstack-poisoned]

guangyey committed:
Update on "Support xpu host allocator"




[ghstack-poisoned]

gujinghui commented:
@pytorchbot rebase

guangyey commented:
@albanD @malfet could you help review this PR?

albanD commented on aten/src/ATen/xpu/CachingHostAllocator.h:
```diff
@@ -0,0 +1,23 @@
+#include <c10/core/Allocator.h>
+#include <c10/xpu/XPUStream.h>
+
+namespace at::xpu {
```
Any reason why we can't have these APIs as generic accelerator APIs?

gujinghui replied:
Yes, we can have generic APIs to replace these C-like APIs.
We are going to submit another PR to unify them, to avoid many cuda modifications in this PR.
Is it OK to you? @albanD

guangyey replied:
These APIs are aligned to CUDA, see https://github.com/pytorch/pytorch/blob/a4c8002ee0714db7d19d98dc0c84108521c5ce6f/aten/src/ATen/cuda/CachingHostAllocator.h#L21-L35

These APIs `record_event`, `empty_cache`, and `allocate` are already defined in [CachingHostAllocatorInterface](https://github.com/pytorch/pytorch/blob/a4c8002ee0714db7d19d98dc0c84108521c5ce6f/aten/src/ATen/core/CachingHostAllocator.h#L352-L378). As similar as device caching allocator, `getCachingHostAllocator` returns a `c10::Allocator*` that doesn't have `record_event` and `empty_cache`.
Do you mean why don't we define `record_event` and `empty_cache` in `c10::Allocator`? If we do this, we can directly can `at::xpu::getCachingHostAllocator().empty_cache()` or `at::cuda::getCachingHostAllocator().record_event()` instead of `at::xpu::CachingHostAllocator_emptyCache()` or `at::cuda::CachingHostAllocator_recordEvent()`, right? Even further, referring to the device allocator, we can call `at::getHostAllocator(device_type).empty_cache()` as a generic method more conveniently.

I think it is a good idea. And it is our final goal. But currently, we plan to implement it step by step. In this PR, we mainly propose to generalize host caching allocator as device-agnostic, according to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), and add support of XPU host allocator sharing the same caching mechanism and code with CUDA.
 - Next, we aim to generalize device allocator (share the code across each backend).
 - Finally, we intend to unify device and host allocator (have `empty_cache` as generic API in `c10::Allocator`).

Meanwhile, there are some open need to be discussed, like:
 - there are at least two types of allocator in PyTorch: caching, and not caching(like async device allocator). Apparently, `empty_cache` should be the caching allocator's method rather than not the caching allocator. Is it reasonable to define `empty_cache` in `c10::Allocator`? Does it seem like `c10::Allocator` is always a caching allocator?
 - record_event(named in host allocator) accepts three parameters as its inputs since host allocator does not split larger allocations into smaller blocks, unlike the caching device allocator. However, record_stream(named in device allocator) accepts two parameters as its inputs. Do we need to unify them or leave them as they are?

In summary, we leave these APIs here aligning to CUDA and make sure this PR change is relatively small. We accept these APIs as generic accelerator APIs if I capture your point. We will unify them following our plan, 1) first generalize device allocator, and then 2) unify them

guangyey replied:
Do you mean to rename `at::xpu::CachingHostAllocator_emptyCache` to `at::xpu::CachingHostAllocator::empty_cache` including CUDA?

albanD replied:
I think the idea is `at::getHostAllocator(device_type).empty_cache()` yes.
With on top of that, `device_type` being optional and defaulting to the current accelerator `at::getAccelerator()` so that most callers don't have to specify the device_type anyways.

And yes we should use `c10::Stream` to handle the stream objects.

guangyey replied:
Got your point.

albanD commented on torch/xpu/__init__.py:
```diff
@@ -398,6 +398,11 @@ def empty_cache() -> None:
         torch._C._xpu_emptyCache()
 
 
+def _host_allocator():
```
Why do you need this one on the python side?

guangyey replied:
To align with CUDA's API [_host_allocator](https://github.com/pytorch/pytorch/blob/eeb626b46a4d606d6900aec8c690feb8a9d6e88b/torch/cuda/memory.py#L54).
Remove it for now if you have any concerns. It is OK to add back if necessary.

albanD replied:
Yeah lets not add it unless we have a use for it!

guangyey committed:
Update on "Support xpu host allocator"


# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

[ghstack-poisoned]

guangyey committed:
Update on "Support xpu host allocator"


# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

[ghstack-poisoned]

guangyey committed:
Update on "Support xpu host allocator"


# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

[ghstack-poisoned]

albanD reviewed (commented):
Sounds like an ok step to unblock.
We can continue the discussion in the follow up to have a generic API taking the device_type to reduce what we expose from aten::

guangyey commented:
> Sounds like an ok step to unblock. We can continue the discussion in the follow up to have a generic API taking the device_type to reduce what we expose from aten::

Thanks, we will continue to follow this.

guangyey committed:
Update on "Support xpu host allocator"


# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

guangyey committed:
Update on "Support xpu host allocator"


# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

[ghstack-poisoned]

guangyey committed:
Update on "Support xpu host allocator"


# Motivation
This PR mainly covers caching host allocator supported on xpu backend.

# Solution
`XPUCachingHostAllocator` adopts the **same** caching mechanism as cuda via two abstract interfaces -`CachingHostAllocatorImpl` and `CachingHostAllocatorInterface`.

# Additional Context
Following CUDA, this PR adds a new API `getPinnedMemoryAllocator` to support the tensor's memory pinned.

[ghstack-poisoned]
